x/superfluid/keeper/migrate.go

```diff
@@ -0,0 +1,80 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	cl "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity"
+	cltypes "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v14/x/gamm/types"
+)
+
+// UnlockAndMigrate unlocks a balancer pool lock, exits the pool and migrates the LP position to a full range concentrated liquidity position.
+// If the lock is also superfluid delegated, it will undelegate the superfluid position.
+func (k Keeper) UnlockAndMigrate(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (amount0, amount1 sdk.Int, liquidity sdk.Dec, poolIdLeaving, poolIdEntering uint64, frozenUntil time.Time, err error) {
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving = gammtypes.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
+
+	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
+	poolIdEntering, err = k.gk.GetLinkedConcentratedPool(ctx, poolIdLeaving)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, err
+	}
+
+	// Get the concentrated pool from the provided ID and type cast it to ConcentratedPoolExtension.
+	poolI, err := k.clk.GetPool(ctx, poolIdEntering)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, err
+	}
+	concentratedPool, ok := poolI.(cltypes.ConcentratedPoolExtension)
+	if !ok {
+		// If the conversion fails, return an error.
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
+	}
+
+	// Consistency check that lockID corresponds to sender, and contains correct LP shares.
+	// These are expected to be true by the caller, but good to double check
+	lock, err := k.validateLockForUnpool(ctx, sender, poolIdLeaving, lockId)
```

Comment:
How do we compare the correct LP shares part? That is, how do we know that the lock corresponds to the right amount of shares?

Reply:
I think we need a check and a test that `sharesToMigrate` and lock's shares are the same

Reply:
CL shares and balancer shares are not 1:1. With that being said, we test what the shares represent in terms of the underlying which is all that really matters IMO. Let me know if this addresses that concern.

Reply:
Ohh I think I see the confusion now, the comment says "contains correct LP shares" but what it actually means is "contains correct denom LP shares".

Reply:
Added this clarifying comment, lmk if that is sufficient for you: https://github.com/osmosis-labs/osmosis/pull/4129/commits/28f04163b4c7d1bccd2021d44dc209d4f79b9af6

Reply:
I guess my main question is with regards to the number of shares locked being greater than the number of shares we migrate.

Say we migrate 4 shares but the lock contains 10. Then, we exit 4/10 shares. What happens to the other 6? Currently, it seems that they are left unlocked.

Maybe we should relock them?

Also, could you please document in the method spec and test the final behavior of this excess lock share amount

Reply:
Woh, this is a great point. Thanks for pointing this out! Let me address it before we merge this

Reply:
@p0mvn I believe the following commit solves the issue you found, thanks for pointing this out!

https://github.com/osmosis-labs/osmosis/pull/4129/commits/68c7991099c48746ef1fb25c640cb7b04e3328af